relayer/src/chain/cosmos/batch.rs

```diff
@@ -0,0 +1,174 @@
+use ibc::events::IbcEvent;
+use ibc_proto::google::protobuf::Any;
+use tendermint_rpc::endpoint::broadcast::tx_sync::Response;
+use tendermint_rpc::{HttpClient, Url};
+use tonic::codegen::http::Uri;
+
+use crate::chain::cosmos::retry::send_tx_with_account_sequence_retry;
+use crate::chain::cosmos::types::account::Account;
+use crate::chain::cosmos::types::tx::TxSyncResult;
+use crate::chain::cosmos::wait::wait_for_block_commits;
+use crate::config::types::Memo;
+use crate::config::ChainConfig;
+use crate::error::Error;
+use crate::keyring::KeyEntry;
+
+pub async fn send_batched_messages_and_wait_commit(
+    config: &ChainConfig,
+    rpc_client: &HttpClient,
+    rpc_address: &Url,
+    grpc_address: &Uri,
+    key_entry: &KeyEntry,
+    account: &mut Account,
+    tx_memo: &Memo,
+    messages: Vec<Any>,
+) -> Result<Vec<IbcEvent>, Error> {
+    if messages.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let mut tx_sync_results = send_messages_as_batches(
+        config,
+        rpc_client,
+        grpc_address,
+        key_entry,
+        account,
+        tx_memo,
+        messages,
+    )
+    .await?;
+
+    wait_for_block_commits(
+        &config.id,
+        rpc_client,
+        rpc_address,
+        &config.rpc_timeout,
+        &mut tx_sync_results,
+    )
+    .await?;
+
+    let events = tx_sync_results
+        .into_iter()
+        .flat_map(|el| el.events)
+        .collect();
+
+    Ok(events)
+}
+
+pub async fn send_batched_messages_and_wait_check_tx(
+    config: &ChainConfig,
+    rpc_client: &HttpClient,
+    grpc_address: &Uri,
+    key_entry: &KeyEntry,
+    account: &mut Account,
+    tx_memo: &Memo,
+    messages: Vec<Any>,
+) -> Result<Vec<Response>, Error> {
+    if messages.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let batches = batch_messages(config, messages)?;
+
+    let mut responses = Vec::new();
+
+    for batch in batches {
+        let response = send_tx_with_account_sequence_retry(
+            config,
+            rpc_client,
+            grpc_address,
+            key_entry,
+            account,
+            tx_memo,
+            batch,
+            0,
+        )
+        .await?;
+
+        responses.push(response);
+    }
+
+    Ok(responses)
+}
+
+async fn send_messages_as_batches(
```

Comment:
This extracts the code from:

https://github.com/informalsystems/ibc-rs/blob/3e34b4b916eb91f680412e5fdd3ea045a99526d2/relayer/src/chain/cosmos.rs#L1095-L1100